### PR TITLE
Optimize reads/queries to Gateway and only send SessionToken when applicable

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
When using TransportMode==Gateway we always send the SessionToken today when the default consistency is Session even when on request level only Eventual consistency is requested. Because of the 16KB header size limitation this can make it impossible to execute queries/readFeed because the combined session token (across al pk ranges) is used. With this change the session token is only sent when appplicable - optimizing when customer opted-for lower consistency and unblocking queries with eventual consistency for accounts with large number of partitions.

### CODE REVIEW

**Comment by FabianMeiswinkel** [comment]
Same issue exists in .Net btw. - will port the change to .Net after check-in as well.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java [hunk 1]
```diff
@@ -71,6 +75,31 @@ CosmosQueryRequestOptions setPartitionKeyRangeIdInternal(String partitionKeyRang
...
+     * Sets the consistency level required for the request. The effective consistency level
+     * can only be reduce for read/query requests. So when the Account's default consistency level
+     * is for example Session you can specify on a request-by-request level for individual requests
+     * that Eventual consistency is sufficient - which could reduce the latency and RU charges for this
+     * request but will not guarantee session consistency (read-your-own-write) anymore
+     *
+     * @param consistencyLevel the consistency level.
+     * @return the CosmosItemRequestOptions.
+     */
+    public CosmosQueryRequestOptions setConsistencyLevel(ConsistencyLevel consistencyLevel) {
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:98]: Are we now supporting request level consistency now?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:98]: We have always been supporting it in .Net - In Java V4 I made it public for read/query where it most commonly used my customers

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DocumentQueryExecutionContextBase.java [hunk 2]
```diff
@@ -131,11 +131,23 @@ public CosmosQueryRequestOptions getFeedOptions(String continuationToken, Intege
         Map<String, String> requestHeaders = new HashMap<>();
 
         ConsistencyLevel defaultConsistencyLevel = this.client.getDefaultConsistencyLevelAsync();
-        ConsistencyLevel desiredConsistencyLevel = this.client.getDesiredConsistencyLevelAsync();
+        ConsistencyLevel desiredConsistencyLevel = cosmosQueryRequestOptions.getConsistencyLevel() != null ?
+            cosmosQueryRequestOptions.getConsistencyLevel():
+            this.client.getDesiredConsistencyLevelAsync();
+
+        boolean sessionTokenApplicable =
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DocumentQueryExecutionContextBase.java:138]: What are session-container semantics with request level overrides? 

ex: With eventual and session SxS
- Is session expected to progress even with eventual requests context?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java [hunk 3]
```diff
@@ -431,21 +431,28 @@ private void applySessionToken(RxDocumentServiceRequest request) {
         Map<String, String> headers = request.getHeaders();
         Objects.requireNonNull(headers, "RxDocumentServiceRequest::headers is required and cannot be null");
 
+        String requestConsistencyLevel = headers.get(HttpConstants.HttpHeaders.CONSISTENCY_LEVEL);
+
+        boolean sessionTokenApplicable =
+            Strings.areEqual(requestConsistencyLevel, ConsistencyLevel.SESSION.toString()) ||
```
Comment (j82w) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java:437]: Should it ignore case?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java [hunk 4]
```diff
@@ -431,21 +431,28 @@ private void applySessionToken(RxDocumentServiceRequest request) {
...
+                (this.defaultConsistencyLevel == ConsistencyLevel.SESSION &&
+                    // skip applying the session token when Eventual Consistency is explicitly requested
+                    // on request-level for data plane operations.
+                    // The session token is ignored on teh backend/gateway in this case anyway
+                    // and the session token can be rather large (even run in the 16 KB header length problem
+                    // on the gateway - so not worth sending when not needed
+                    (!request.isReadOnlyRequest() ||
+                        request.getResourceType() != ResourceType.Document ||
```
Comment (j82w) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java:445]: 1. Should execute stored procedure be included?
2. What about get query plan? That never needs the session token.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DocumentQueryExecutionContextBase.java (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DocumentQueryExecutionContextBase.java:138]: It is independent of the change here. Backend ignores the SessionToken when we request Eventual Consistency on a request level. The response still contains a session token - even when request is for eventual without session token.  We update the SessionContainer with the session token from the response (if it is newer). So session would progress even for individual eventual consistency requests.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java:437]: Yes - thanks - good catch .

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java (hunk 4 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java:445]: Query plan - good point. Stored Procedure I would not want to touch right now. Semantics aren't clear enough for me - and I am not aware of a single scenario where customers would want to adjust consistency level for SPs individually.

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java:445]: Fixed in next iteration

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosItemRequestOptions.java [hunk 5]
```diff
@@ -101,17 +101,21 @@ public CosmosItemRequestOptions setIfNoneMatchETag(String ifNoneMatchETag) {
...
         return consistencyLevel;
     }
 
     /**
-     * Sets the consistency level required for the request.
+     * Sets the consistency level required for the request. The effective consistency level
+     * can only be reduce for read/query requests. So when the Account's default consistency level
+     * is for example Session you can specify on a request-by-request level for individual requests
+     * that Eventual consistency is sufficient - which could reduce the latency and RU charges for this
+     * request but will not guarantee session consistency (read-your-own-write) anymore
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosItemRequestOptions.java:113]: If the account default consistency is session, client default consistency is set to be eventual, but for an individual request consistency is set to session, we will have a problem.

By default in java v4 we don't capture session tokens for non session consistency.
we should document that if the user wants to use this per request option, the user has to enable per request session capturing on the cosmos-client-builder and we should make the method for that public.

see `CosmosClientBuilder:sessionCapturingOverrideEnabled()`

`CosmosItemRequestOption.setConsistencyLevel()` and `CosmosClientBuilder:sessionCapturingOverrideEnabled()` should go hand in hand.

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java [hunk 6]
```diff
@@ -101,6 +102,30 @@ public void readItem() throws Exception {
...
+            idAndPkValue,
+            new PartitionKey(idAndPkValue),
+            new CosmosItemRequestOptions()
+                // generate an invalid session token large enough to cause an error in Gateway
+                // due to header being too long
+                .setSessionToken(StringUtils.repeat("SomeManualInvalidSessionToken", 2000))
+                .setConsistencyLevel(ConsistencyLevel.EVENTUAL),
+            ObjectNode.class);
+
+        System.out.println("REQUEST DIAGNOSTICS: " + readResponse1.getDiagnostics().toString());
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java:125]:
```suggestion
        logger.info("REQUEST DIAGNOSTICS: {}" readResponse1.getDiagnostics());
```

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java [hunk 7]
```diff
@@ -175,6 +199,41 @@ public void queryItems() throws Exception{
...
+            new CosmosQueryRequestOptions()
+                // generate an invalid session token large enough to cause an error in Gateway
+                // due to header being too long
+                .setSessionToken(StringUtils.repeat("SomeManualInvalidSessionToken", 2000))
+                .setConsistencyLevel(ConsistencyLevel.EVENTUAL);
+
+        CosmosPagedIterable<ObjectNode> feedResponseIterator1 =
+            container.queryItems(query, cosmosQueryRequestOptions, ObjectNode.class);
+        feedResponseIterator1.handle((r) -> System.out.println("Query RequestDiagnostics: " + r.getCosmosDiagnostics().toString()));
+        System.out.println("hello world");
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java:223]: please remove.

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java [hunk 8]
```diff
@@ -175,6 +199,41 @@ public void queryItems() throws Exception{
...
+
+        // Very basic validation
+        assertThat(feedResponseIterator1.iterator().hasNext()).isTrue();
+        assertThat(feedResponseIterator1.stream().count() == 1);
+
+        SqlQuerySpec querySpec = new SqlQuerySpec(query);
+        CosmosPagedIterable<ObjectNode> feedResponseIterator3 =
+            container.queryItems(querySpec, cosmosQueryRequestOptions, ObjectNode.class);
+        feedResponseIterator3.handle((r) -> System.out.println("Query RequestDiagnostics: " + r.getCosmosDiagnostics().toString()));
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java:232]: ditto on the System.out.println -> logger

**Review by moderakh — CHANGES_REQUESTED** [review]
If the account default consistency is session, client default consistency is set to be eventual, but for an individual request consistency is set to session, we need tuning; otherwise session consistency will not be met.

By default in java v4 we don't capture session tokens for non session consistency.
we should document that if the user wants to use this per request option, the user has to enable per request session capturing on the cosmos-client-builder and we should make the method for that public.

see `CosmosClientBuilder:sessionCapturingOverrideEnabled()`

`CosmosItemRequestOption.setConsistencyLevel()` and `CosmosClientBuilder:sessionCapturingOverrideEnabled()` should go hand in hand.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java [hunk 9]
```diff
@@ -114,6 +114,11 @@ public RxGatewayStoreModel(
     private Mono<RxDocumentServiceResponse> query(RxDocumentServiceRequest request) {
         if(request.getOperationType() != OperationType.QueryPlan) {
             request.getHeaders().put(HttpConstants.HttpHeaders.IS_QUERY, "true");
+        } else {
+            // Session token is irrelevant for query plan
+            // Gateway can hit header limitation issues quickly so removing it here
+            // to protect accounts with large number of partitions (big session tokens)
+            request.getHeaders().remove(HttpConstants.HttpHeaders.SESSION_TOKEN);
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java:121]: Removed the code

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java [hunk 10]
```diff
@@ -431,21 +438,28 @@ private void applySessionToken(RxDocumentServiceRequest request) {
...
+            Strings.areEqualIgnoreCase(requestConsistencyLevel, ConsistencyLevel.SESSION.toString()) ||
+                (this.defaultConsistencyLevel == ConsistencyLevel.SESSION &&
+                    // skip applying the session token when Eventual Consistency is explicitly requested
+                    // on request-level for data plane operations.
+                    // The session token is ignored on teh backend/gateway in this case anyway
+                    // and the session token can be rather large (even run in the 16 KB header length problem
+                    // on the gateway - so not worth sending when not needed
+                    (!request.isReadOnlyRequest() ||
+                        request.getResourceType() != ResourceType.Document ||
+                        !Strings.areEqualIgnoreCase(requestConsistencyLevel, ConsistencyLevel.EVENTUAL.toString())));
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java:453]: ACK

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java [hunk 11]
```diff
@@ -431,21 +438,28 @@ private void applySessionToken(RxDocumentServiceRequest request) {
         Map<String, String> headers = request.getHeaders();
         Objects.requireNonNull(headers, "RxDocumentServiceRequest::headers is required and cannot be null");
 
+        String requestConsistencyLevel = headers.get(HttpConstants.HttpHeaders.CONSISTENCY_LEVEL);
+
+        boolean sessionTokenApplicable =
+            Strings.areEqualIgnoreCase(requestConsistencyLevel, ConsistencyLevel.SESSION.toString()) ||
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java:444]: Fixed

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java (hunk 6 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java:125]: Fixed

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java (hunk 7 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java:223]: Fixed

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java (hunk 8 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java:232]: Fixed

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosItemRequestOptions.java (hunk 5 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosItemRequestOptions.java:113]: Nice catch - thanks. Added documentation and made sessionCapturingOverrideEnabled public

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosItemRequestOptions.java [hunk 12]
```diff
@@ -100,18 +100,25 @@ public CosmosItemRequestOptions setIfNoneMatchETag(String ifNoneMatchETag) {
...
+     * NOTE: If the consistency-level set on a request level here is higher than the default consistency
+     * level set when constructing the CosmosClient instance via CosmosClientBuilder.consistencyLevel
+     * session token capturing also needs to be enabled by calling
+     * CosmosClientBuilder:sessionCapturingOverrideEnabled(true).
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosItemRequestOptions.java:116]: `CosmosClientBuilder:sessionCapturingOverrideEnabled(true)` would only be needed when you set the request consistency to Session.

This is not needed if you set the request consistency to Strong nor for Bounded-Staleness.

**Review by moderakh — APPROVED** [review]
LGTM.

**Pull request merged**